### DISCUSSION TITLE
An SWC enabled solution for Inline SVGs

### DISCUSSION THREAD

**Discussion by boennemann** [vercel/next.js#33161] [Help]
Turning this into a separate discussion from: https://github.com/vercel/next.js/discussions/30174#discussioncomment-1900852

Seeing lots of upvotes, but I'm not sure it gets in front of people who could help.

> I feel like I am missing something very obvious, but I can't find out what, so please excuse me asking:
> We're down to inline-react-svg being our last babel plugin. Inlining SVGs is necessary to modify the SVGs with CSS/props as far as I know. It's very important for us to use SVGs for icons/interface components, and to be able to pretty much just copy them from Figma.
> Now I only see a few people in [here](https://github.com/vercel/next.js/discussions/30174) using this plugin. I don't see any mention on the official Next.js docs how to use SVGs, other than `<Image src="/icon.svg" />`.
It's also not mentioned in the [original post](https://github.com/vercel/next.js/discussions/30174) as something being on the roadmap for the swc based compiler.
> That leaves me wondering, whether we're doing something completely outdated. How are you dealing with icons, if not SVGs? Is this really not covered by Next.js (without babel/webpack) somewhere?
How would we make this work with swc?
> The only option I see right now is converting the svgs to jsx and then inlining them into the source code. Is that really the way to go?
> Helpful for any hint. We really want to move to swc 🤞

**Comment by kylemh** [comment]
see https://github.com/vercel/next.js/discussions/26167

**Comment by OscarBarrett** [comment]
Using [@svgr/webpack](https://www.npmjs.com/package/@svgr/webpack) and the following config works without babel:

```js
config.module.rules.push({
  test: /\.svg$/,
  issuer: /\.tsx?$/,
  include: [options.dir],
  use: [
    'next-swc-loader',
    {
      loader: '@svgr/webpack',
      options: { babel: false }
    }
  ],
});
```

In your component:

```tsx
import Image from './image.svg';

....
<Image />
```

In a skeleton app with one svg import I'm not seeing a performance hit to builds. Not sure how well it scales.

**Reply by boennemann** [reply]
Wait a second, could this be the misunderstanding I had all along.

Is swc replacing babel + webpack in Next.js, or only babel?

> SWC can be used for compilation, minification, bundling, and more

The docs only mention _can_ with regards to the bundling part. But is it?

**Reply by kylemh** [reply]
Only babel (and terser)

**Reply by alimertcakar** [reply]
Did the same, I'm getting a "You may need an additional loader to handle the result of these loaders." error, what might be the problem? @OscarBarrett

**Reply by OscarBarrett** [reply]
My guess would be you were previously inlining svgs from a location not covered under `options.dir`.

For example, if you have a component library that includes svgs that you wish to inline, (i.e. a path you were adding to `config.resolve.modules` and/or `config.resolve.alias`) then you should also add that to the `include` parameter of the rule.

The path after the `error` log should indicate whether this is the case.

**Reply by elliott-w** [reply]
This no longer works in Next.js v13 (I'm on 13.0.2), getting:

`TypeError: The "from" argument must be of type string. Received undefined`

@kylemh any ideas why?

My current workaround is setting the config to:

```js
config.module.rules.push({
  test: /\.svg$/,
  issuer: /\.tsx?$/,
  include: [options.dir],
  use: ['@svgr/webpack']
})
```

But I'm fairly sure it now uses babel to compile those `.tsx` files that contain `.svg` imports rather than swc.

**Reply by kylemh** [reply]
I never tried using `next-swc-loader` with `@svgr/webpack`. 🤷‍♂️

**Reply by fakundo** [reply]
> This no longer works in Next.js v13 (I'm on 13.0.2), getting:
> 
> `TypeError: The "from" argument must be of type string. Received undefined`

Seems I found a workaround by using defaultLoaders from options.

```ts
webpack: (config, options) => {
  config.module.rules.push(
    {
      test: /\.svg$/,
      use: [
        options.defaultLoaders.babel,
        {
          loader: '@svgr/webpack',
          options: { babel: false },
        },
      ],
    },
  );
  return config;
}
```

For me `options.defaultLoaders.babel` actually returns configured `next-swc-loader`.

**Reply by vincentvella** [reply]
I was still having some issues with the above so I created a plugin that's easy to use with next-compose-plugins:
```js
const withSVGR = (nextConfig) => {
  nextConfig.webpack = (config, options) => {
    config.module.rules.map((rule) => {
      if (rule.test !== undefined && rule.test.source.includes('|svg|')) {
        rule.test = new RegExp(rule.test.source.replace('|svg|', '|'));
      }
    });
    config.module.rules.push({
      test: /\.svg$/i,
      issuer: /\.[jt]sx?$/,
      use: ['@svgr/webpack'],
    });
    return config;
  };
  return nextConfig;
};
```

**Reply by LeVuMinhHuy** [reply]
you are my hero @fakundo  !

**Comment by rtatarinov** [comment]
@fakundo @OscarBarrett @elliott-w it seems any of solution doesn't work with `13.4.3`. When I use `next-swc-loader` I still have `TypeError: The "from" argument must be of type string. Received undefined`. Any other config returns `You may need an appropriate loader to handle this file type, currently no loaders are configured to process this file. See https://webpack.js.org/concepts#loaders`

Any workaround to fix it?

**Comment by matepapp** [comment]
has anyone figured out how to handle SVGR with Next.js v14 using SWC? I've tried every possible combination without any success so far

**Reply by dymoo** [reply]
> has anyone figured out how to handle SVGR with Next.js v14 using SWC? I've tried every possible combination without any success so far

Adding 

```typescript
experimental: {
  forceSwcTransforms: true
}
```

and 

```typescript
webpack: (config, options) => {
  config.module.rules.push({
    test: /\.svg$/,
    use: [
      options.defaultLoaders.babel,
      {
        loader: "@svgr/webpack",
        options: { babel: false },
      },
    ],
  });
  return config;
},
```

to my next config seems to work fine... 'forceSwcTransforms' should do just that ..

**Comment by cdierkens** [comment]
The year is 2025 and there isn't an inline SVG solution for NextJS. I'm really not shocked.Par for the corse for NextJS. No, I'm not just gonna pass it to next/image, and no I'm not gonna use a webpack config. It's 2025.

**Reply by enBonnet** [reply]
> The year is 2025 and there isn't an inline SVG solution for NextJS. I'm really not shocked.Par for the corse for NextJS. No, I'm not just gonna pass it to next/image, and no I'm not gonna use a webpack config. It's 2025.

Between many languages, you decide to write in facts.

**Comment by vitalets** [comment]
I've build a [custom turbopack loader](https://github.com/vitalets/turbopack-inline-svg-loader) that conditionally inlines SVGs as data URIs, based on the file size. It works in Next.js 16+, that introduced turbopack *condition rules*.

Configuration:
```js
const nextConfig = {
  turbopack: {
    rules: {
      '*.svg': {
        loaders: ['turbopack-inline-svg-loader'],
        condition: {
          content: /^[\s\S]{0,4000}$/, // <-- Inline SVGs smaller than ~4Kb (since Next.js v16)
        },
        as: '*.js',
      },
    },
  },
  // ...
};
```
Usage:
```tsx
import Image from 'next/image';
import myIcon from './icon.svg';

export default function Page() {
  return <Image src={myIcon} alt="my icon" />;
}
```

Rendered output:

<img width="643" height="40" alt="image" src="https://github.com/user-attachments/assets/ead9cc27-78c1-458a-96de-ab5e0ea0d8bc" />
